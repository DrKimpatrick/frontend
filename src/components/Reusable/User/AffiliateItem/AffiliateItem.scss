@import '../UserItem/UserItem.scss';

.affiliateContainer {
  border: none !important;
  .userItemList {
    min-height: 0px !important;
  }
  .affiliateItems {
    border: 2px solid #dadada5e;
    border-radius: 5px;
  }
}
