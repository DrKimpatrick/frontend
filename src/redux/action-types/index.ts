// This is where we will add action types