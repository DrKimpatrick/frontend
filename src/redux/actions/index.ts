// This is where we will add our action creators
export {};
