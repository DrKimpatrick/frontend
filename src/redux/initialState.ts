import users from 'redux/initialStates/users';

const initialState = {
  users
};

export default initialState;
