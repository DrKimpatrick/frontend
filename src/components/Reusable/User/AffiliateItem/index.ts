import AffiliateItem from './AffiliateItem';

export { AffiliateItem };

export * from './AffiliateItem';
