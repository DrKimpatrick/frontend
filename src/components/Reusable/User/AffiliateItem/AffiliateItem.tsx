import React, { FC } from 'react';
import './AffiliateItem.scss';
import ReactPaginate from 'react-paginate';
import { Course, CourseStatus } from 'redux/action-types/course';
import { ArrowDropDown } from '@material-ui/icons';
import { changeCourseStatus } from 'redux/actions/course';
import { useDispatch } from 'react-redux';
import { NoItemFound } from 'components/Reusable';
import { OnPageChangeCallback } from '..';

const options = [
  { label: CourseStatus.Accepted, value: CourseStatus.Accepted },
  { label: CourseStatus.Pending, value: CourseStatus.Pending },
  { label: CourseStatus.Declined, value: CourseStatus.Declined }
];

export interface ChangeAffiliateStatus {
  course: Course;
  status: string;
  pageCount: number;
}

interface Props {
  items: Course[];
  totalItems: number;
  currentPage: number;
  itemPerPage: number;
  pageChange: OnPageChangeCallback;
  setPage: (value: number) => void;
  setOffset: (value: number) => void;
  offset: number;
  type: string;
}

const AffiliateItem: FC<Props> = props => {
  const {
    items,
    totalItems,
    currentPage,
    itemPerPage,
    pageChange,
    setOffset,
    setPage,
    offset,
    type
  } = props;

  const pageCount = Math.ceil(totalItems / itemPerPage);

  const dispatch = useDispatch();

  const changeStatus = (values: ChangeAffiliateStatus) => {
    const { course, status } = values;

    if (items.length === 1) {
      const newPage = Math.ceil(currentPage - 1);
      const newOffset = Math.ceil(newPage * itemPerPage);
      setPage(newPage);
      setOffset(offset === 0 ? 0 : newOffset);
      changeCourseStatus({
        offset: offset === 0 ? 0 : newOffset,
        limit: itemPerPage,
        status,
        course,
        type
      })(dispatch);
    } else {
      changeCourseStatus({ offset, limit: itemPerPage, status, course, type })(
        dispatch
      );
    }
  };

  return (
    <>
      <div className="userContainer affiliateContainer">
        {items && items.length > 0 && (
          <div className="userItemList">
            <ul className="userItems affiliateItems">
              {items.map((item, index) => (
                <li key={index}>
                  <div className="image">
                    <img src={item.coverImageLink} alt="" />
                  </div>
                  <div className="details">
                    <h1>{item.name}</h1>
                    <h6>{item.userId.username}</h6>
                    <button
                      type="button"
                      onClick={() =>
                        window.open(item.existingCourseLink, '_blank')
                      }
                    >
                      View
                    </button>
                  </div>
                  <div className="action">
                    <h5>Change Status</h5>
                    <div className="customSelect">
                      <select
                        name="options"
                        value={item.verificationStatus}
                        onChange={e =>
                          changeStatus({
                            status: e.target.value,
                            course: item,
                            pageCount
                          })
                        }
                      >
                        <option value={item.verificationStatus}>
                          {item.verificationStatus}
                        </option>
                        {options
                          .filter(opt => opt.value !== item.verificationStatus)
                          .map((opt, i) => (
                            <option key={i} value={opt.value}>
                              {opt.label}
                            </option>
                          ))}
                      </select>
                      <span className="customArrow">
                        <ArrowDropDown />
                      </span>
                    </div>
                  </div>
                </li>
              ))}
              {pageCount > 1 && (
                <li className="paginationSection">
                  <ReactPaginate
                    pageRangeDisplayed={5}
                    previousLabel={'<'}
                    nextLabel={'>'}
                    pageCount={pageCount}
                    onPageChange={pageChange}
                    marginPagesDisplayed={2}
                    breakLabel="..."
                    forcePage={currentPage}
                    activeClassName="active"
                  />
                </li>
              )}
            </ul>
          </div>
        )}
      </div>
      {items && items.length <= 0 && <NoItemFound />}
    </>
  );
};

export default AffiliateItem;
