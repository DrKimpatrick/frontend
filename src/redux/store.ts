import { applyMiddleware, createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import apiMiddleware from 'middleware/apiMiddleware';
import rootReducer from './reducers/rootReducer';
import initialState from './initialState';

const middleware = [apiMiddleware];

const store = createStore(
  rootReducer,
  initialState,
  composeWithDevTools(applyMiddleware(...middleware))
);

export type RootState = ReturnType<typeof rootReducer>;

export default store;
