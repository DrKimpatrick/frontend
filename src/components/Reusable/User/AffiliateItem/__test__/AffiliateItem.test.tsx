import React from 'react';
import ReactDom from 'react-dom';
import { Provider } from 'react-redux';
import { Store } from 'redux';
import { BrowserRouter as Router } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import thunk from 'redux-thunk';
import { cleanup, render, fireEvent } from '@testing-library/react';
import AffiliateItem from '../AffiliateItem';

let store: Store;

let middleware = [thunk];

let mockStore = configureMockStore(middleware);

const mockSetOffset = jest.fn();

const mockSetPage = jest.fn();

const mockDispatch = jest.fn();

const props = {
  currentPage: 1,
  setOffset: mockSetOffset,
  setPage: mockSetPage,
  items: [
    {
      _id: 'courseId',
      name: 'courseName',
      userId: {
        username: 'username',
        email: 'email@gmail.com',
        avatar: 'avatar.png'
      },
      existingCourseLink: 'https://link.com',
      coverImageLink: 'https://coverImage.com/image.png',
      verificationStatus: 'pending',
      instructor: 'instructor',
      languageTaught: 'english',
      createdAt: new Date().toDateString(),
      updatedAt: new Date().toDateString(),
      currentLangSpecsUpdated: true
    }
  ],
  totalItems: 1,
  itemPerPage: 2,
  pageChange: jest.fn(),
  offset: 0,
  type: 'CourseTypes'
};

describe('AffiliateItem', () => {
  beforeEach(() => {
    store = mockStore({});
  });

  afterEach(cleanup);

  jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useDispatch: mockDispatch
  }));

  it('should render without crashing', () => {
    const div = document.createElement('div');

    ReactDom.render(
      <Provider store={store}>
        <AffiliateItem {...props} />
      </Provider>,
      div
    );
  });

  it('should create snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <AffiliateItem {...props} />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe('User should change status', () => {
    it('should allow user to change status', () => {
      const { container } = render(
        <Provider store={store}>
          <AffiliateItem {...props} />
        </Provider>
      );

      const selectMenu: HTMLSelectElement | any = container.querySelector(
        'select'
      );

      fireEvent.change(selectMenu, () => ({
        target: {
          value: 'love'
        }
      }));

      expect(mockDispatch).toBeTruthy();
    });

    it('should allow user to change even when items is more than one', () => {
      const newCourses = [props.items[0], props.items[0]];

      const newProps = {
        ...props,
        items: newCourses
      };
      const { container } = render(
        <Provider store={store}>
          <Router>
            <AffiliateItem {...newProps} />
          </Router>
        </Provider>
      );

      const selectMenu: HTMLSelectElement | any = container.querySelector(
        'select'
      );

      fireEvent.change(selectMenu, () => ({
        target: {
          value: 'love'
        }
      }));

      expect(mockDispatch).toBeTruthy();
    });
  });

  describe('render not found component', () => {
    it('should render not found component', () => {
      const newProps = {
        ...props,
        items: []
      };
      const { container } = render(
        <Provider store={store}>
          <Router>
            <AffiliateItem {...newProps} />
          </Router>
        </Provider>
      );

      const div: Element | any = container.querySelector('.message');

      expect(div).toHaveTextContent('No result found');
    });
  });
});
