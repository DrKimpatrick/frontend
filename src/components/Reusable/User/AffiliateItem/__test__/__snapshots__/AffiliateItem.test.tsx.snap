// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AffiliateItem should create snapshot 1`] = `
<div
  className="userContainer affiliateContainer"
>
  <div
    className="userItemList"
  >
    <ul
      className="userItems affiliateItems"
    >
      <li>
        <div
          className="image"
        >
          <img
            alt=""
            src="https://coverImage.com/image.png"
          />
        </div>
        <div
          className="details"
        >
          <h1>
            courseName
          </h1>
          <h6>
            username
          </h6>
          <button
            onClick={[Function]}
            type="button"
          >
            View
          </button>
        </div>
        <div
          className="action"
        >
          <h5>
            Change Status
          </h5>
          <div
            className="customSelect"
          >
            <select
              name="options"
              onChange={[Function]}
              value="pending"
            >
              <option
                value="pending"
              >
                pending
              </option>
              <option
                value="accepted"
              >
                accepted
              </option>
              <option
                value="declined"
              >
                declined
              </option>
            </select>
            <span
              className="customArrow"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </span>
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
